File: lib/models/user_health_profile.dart
class UserHealthProfile {
  final String? id;
  final String userId;
  final String bloodGroup;
  final Height? height;
  final List<MedicalCondition> medicalConditions;
  final List<Allergy> allergies;
  final List<Medication> currentMedications;
  final List<HealthGoal> healthGoals;
  final List<EmergencyContact> emergencyContacts;
  final List<FamilyMedicalHistory> familyMedicalHistory;
  final Lifestyle lifestyle;
  final Insurance? insurance;

  UserHealthProfile({
    this.id,
    required this.userId,
    this.bloodGroup = 'Unknown',
    this.height,
    this.medicalConditions = const [],
    this.allergies = const [],
    this.currentMedications = const [],
    this.healthGoals = const [],
    this.emergencyContacts = const [],
    this.familyMedicalHistory = const [],
    required this.lifestyle,
    this.insurance,
  });

  factory UserHealthProfile.fromJson(Map<String, dynamic> json) {
    return UserHealthProfile(
      id: json['_id'],
      userId: json['userId'],
      bloodGroup: json['bloodGroup'] ?? 'Unknown',
      height: json['height'] != null ? Height.fromJson(json['height']) : null,
      medicalConditions: (json['medicalConditions'] as List?)
          ?.map((e) => MedicalCondition.fromJson(e))
          .toList() ?? [],
      allergies: (json['allergies'] as List?)
          ?.map((e) => Allergy.fromJson(e))
          .toList() ?? [],
      currentMedications: (json['currentMedications'] as List?)
          ?.map((e) => Medication.fromJson(e))
          .toList() ?? [],
      healthGoals: (json['healthGoals'] as List?)
          ?.map((e) => HealthGoal.fromJson(e))
          .toList() ?? [],
      emergencyContacts: (json['emergencyContacts'] as List?)
          ?.map((e) => EmergencyContact.fromJson(e))
          .toList() ?? [],
      familyMedicalHistory: (json['familyMedicalHistory'] as List?)
          ?.map((e) => FamilyMedicalHistory.fromJson(e))
          .toList() ?? [],
      lifestyle: Lifestyle.fromJson(json['lifestyle'] ?? {}),
      insurance: json['insurance'] != null ? Insurance.fromJson(json['insurance']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'userId': userId,
      'bloodGroup': bloodGroup,
      if (height != null) 'height': height!.toJson(),
      'medicalConditions': medicalConditions.map((e) => e.toJson()).toList(),
      'allergies': allergies.map((e) => e.toJson()).toList(),
      'currentMedications': currentMedications.map((e) => e.toJson()).toList(),
      'healthGoals': healthGoals.map((e) => e.toJson()).toList(),
      'emergencyContacts': emergencyContacts.map((e) => e.toJson()).toList(),
      'familyMedicalHistory': familyMedicalHistory.map((e) => e.toJson()).toList(),
      'lifestyle': lifestyle.toJson(),
      if (insurance != null) 'insurance': insurance!.toJson(),
    };
  }

  List<Medication> getActiveMedications() {
    return currentMedications.where((med) => med.isActive).toList();
  }

  List<MedicalCondition> getActiveConditions() {
    return medicalConditions.where((cond) => cond.isActive).toList();
  }

  EmergencyContact? getPrimaryEmergencyContact() {
    return emergencyContacts.firstWhere(
      (contact) => contact.isPrimary,
      orElse: () => emergencyContacts.isNotEmpty ? emergencyContacts[0] : EmergencyContact(name: '', phone: '', relation: ''),
    );
  }
}

class Height {
  final double? value;
  final String unit;

  Height({
    this.value,
    this.unit = 'cm',
  });

  factory Height.fromJson(Map<String, dynamic> json) {
    return Height(
      value: json['value']?.toDouble(),
      unit: json['unit'] ?? 'cm',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (value != null) 'value': value,
      'unit': unit,
    };
  }
}

class MedicalCondition {
  final String? id;
  final String name;
  final DateTime? diagnosedDate;
  final String? notes;
  final bool isActive;

  MedicalCondition({
    this.id,
    required this.name,
    this.diagnosedDate,
    this.notes,
    this.isActive = true,
  });

  factory MedicalCondition.fromJson(Map<String, dynamic> json) {
    return MedicalCondition(
      id: json['_id'],
      name: json['name'],
      diagnosedDate: json['diagnosedDate'] != null ? DateTime.parse(json['diagnosedDate']) : null,
      notes: json['notes'],
      isActive: json['isActive'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'name': name,
      if (diagnosedDate != null) 'diagnosedDate': diagnosedDate!.toIso8601String(),
      if (notes != null) 'notes': notes,
      'isActive': isActive,
    };
  }
}

class Allergy {
  final String? id;
  final String allergen;
  final String type;
  final String severity;
  final String? reaction;
  final String? notes;

  Allergy({
    this.id,
    required this.allergen,
    this.type = 'Other',
    this.severity = 'Moderate',
    this.reaction,
    this.notes,
  });

  factory Allergy.fromJson(Map<String, dynamic> json) {
    return Allergy(
      id: json['_id'],
      allergen: json['allergen'],
      type: json['type'] ?? 'Other',
      severity: json['severity'] ?? 'Moderate',
      reaction: json['reaction'],
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'allergen': allergen,
      'type': type,
      'severity': severity,
      if (reaction != null) 'reaction': reaction,
      if (notes != null) 'notes': notes,
    };
  }
}

class Medication {
  final String? id;
  final String name;
  final String? dosage;
  final String? frequency;
  final DateTime? startDate;
  final DateTime? endDate;
  final String? prescribedBy;
  final String? purpose;
  final bool isActive;

  Medication({
    this.id,
    required this.name,
    this.dosage,
    this.frequency,
    this.startDate,
    this.endDate,
    this.prescribedBy,
    this.purpose,
    this.isActive = true,
  });

  factory Medication.fromJson(Map<String, dynamic> json) {
    return Medication(
      id: json['_id'],
      name: json['name'],
      dosage: json['dosage'],
      frequency: json['frequency'],
      startDate: json['startDate'] != null ? DateTime.parse(json['startDate']) : null,
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
      prescribedBy: json['prescribedBy'],
      purpose: json['purpose'],
      isActive: json['isActive'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'name': name,
      if (dosage != null) 'dosage': dosage,
      if (frequency != null) 'frequency': frequency,
      if (startDate != null) 'startDate': startDate!.toIso8601String(),
      if (endDate != null) 'endDate': endDate!.toIso8601String(),
      if (prescribedBy != null) 'prescribedBy': prescribedBy,
      if (purpose != null) 'purpose': purpose,
      'isActive': isActive,
    };
  }
}

class HealthGoal {
  final String? id;
  final String goal;
  final DateTime? targetDate;
  final String status;
  final String? notes;

  HealthGoal({
    this.id,
    required this.goal,
    this.targetDate,
    this.status = 'Not Started',
    this.notes,
  });

  factory HealthGoal.fromJson(Map<String, dynamic> json) {
    return HealthGoal(
      id: json['_id'],
      goal: json['goal'],
      targetDate: json['targetDate'] != null ? DateTime.parse(json['targetDate']) : null,
      status: json['status'] ?? 'Not Started',
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'goal': goal,
      if (targetDate != null) 'targetDate': targetDate!.toIso8601String(),
      'status': status,
      if (notes != null) 'notes': notes,
    };
  }
}

class EmergencyContact {
  final String? id;
  final String name;
  final String phone;
  final String relation;
  final bool isPrimary;

  EmergencyContact({
    this.id,
    required this.name,
    required this.phone,
    required this.relation,
    this.isPrimary = false,
  });

  factory EmergencyContact.fromJson(Map<String, dynamic> json) {
    return EmergencyContact(
      id: json['_id'],
      name: json['name'],
      phone: json['phone'],
      relation: json['relation'],
      isPrimary: json['isPrimary'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'name': name,
      'phone': phone,
      'relation': relation,
      'isPrimary': isPrimary,
    };
  }
}

class FamilyMedicalHistory {
  final String? id;
  final String? relation;
  final String? condition;
  final String? notes;

  FamilyMedicalHistory({
    this.id,
    this.relation,
    this.condition,
    this.notes,
  });

  factory FamilyMedicalHistory.fromJson(Map<String, dynamic> json) {
    return FamilyMedicalHistory(
      id: json['_id'],
      relation: json['relation'],
      condition: json['condition'],
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      if (relation != null) 'relation': relation,
      if (condition != null) 'condition': condition,
      if (notes != null) 'notes': notes,
    };
  }
}

class Lifestyle {
  final String smokingStatus;
  final String alcoholConsumption;
  final String exerciseFrequency;
  final String dietType;

  Lifestyle({
    this.smokingStatus = 'Unknown',
    this.alcoholConsumption = 'Unknown',
    this.exerciseFrequency = 'Unknown',
    this.dietType = 'Unknown',
  });

  factory Lifestyle.fromJson(Map<String, dynamic> json) {
    return Lifestyle(
      smokingStatus: json['smokingStatus'] ?? 'Unknown',
      alcoholConsumption: json['alcoholConsumption'] ?? 'Unknown',
      exerciseFrequency: json['exerciseFrequency'] ?? 'Unknown',
      dietType: json['dietType'] ?? 'Unknown',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'smokingStatus': smokingStatus,
      'alcoholConsumption': alcoholConsumption,
      'exerciseFrequency': exerciseFrequency,
      'dietType': dietType,
    };
  }
}

class Insurance {
  final String? provider;
  final String? policyNumber;
  final DateTime? expiryDate;

  Insurance({
    this.provider,
    this.policyNumber,
    this.expiryDate,
  });

  factory Insurance.fromJson(Map<String, dynamic> json) {
    return Insurance(
      provider: json['provider'],
      policyNumber: json['policyNumber'],
      expiryDate: json['expiryDate'] != null ? DateTime.parse(json['expiryDate']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (provider != null) 'provider': provider,
      if (policyNumber != null) 'policyNumber': policyNumber,
      if (expiryDate != null) 'expiryDate': expiryDate!.toIso8601String(),
    };
  }
}

File: icare-backend/models/UserHealthProfile.js
const mongoose = require('mongoose');

const userHealthProfileSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      unique: true,
      index: true,
    },

    // Basic Health Information
    bloodGroup: {
      type: String,
      enum: ['A+', 'A-', 'B+', 'B-', 'AB+', 'AB-', 'O+', 'O-', 'Unknown'],
      default: 'Unknown',
    },

    height: {
      value: Number,
      unit: {
        type: String,
        enum: ['cm', 'ft'],
        default: 'cm',
      },
    },

    // Medical Conditions
    medicalConditions: [
      {
        name: {
          type: String,
          required: true,
        },
        diagnosedDate: Date,
        notes: String,
        isActive: {
          type: Boolean,
          default: true,
        },
      },
    ],

    // Allergies
    allergies: [
      {
        allergen: {
          type: String,
          required: true,
        },
        type: {
          type: String,
          enum: ['Food', 'Medicine', 'Environmental', 'Other'],
          default: 'Other',
        },
        severity: {
          type: String,
          enum: ['Mild', 'Moderate', 'Severe'],
          default: 'Moderate',
        },
        reaction: String,
        notes: String,
      },
    ],

    // Current Medications
    currentMedications: [
      {
        name: {
          type: String,
          required: true,
        },
        dosage: String,
        frequency: String, // "Once daily", "Twice daily", "As needed"
        startDate: Date,
        endDate: Date,
        prescribedBy: String, // Doctor name
        purpose: String,
        isActive: {
          type: Boolean,
          default: true,
        },
      },
    ],

    // Health Goals
    healthGoals: [
      {
        goal: {
          type: String,
          required: true,
        },
        targetDate: Date,
        status: {
          type: String,
          enum: ['Not Started', 'In Progress', 'Achieved', 'Abandoned'],
          default: 'Not Started',
        },
        notes: String,
      },
    ],

    // Emergency Contacts
    emergencyContacts: [
      {
        name: {
          type: String,
          required: true,
        },
        phone: {
          type: String,
          required: true,
        },
        relation: {
          type: String,
          required: true,
        },
        isPrimary: {
          type: Boolean,
          default: false,
        },
      },
    ],

    // Family Medical History
    familyMedicalHistory: [
      {
        relation: String, // "Father", "Mother", "Sibling"
        condition: String,
        notes: String,
      },
    ],

    // Lifestyle Information
    lifestyle: {
      smokingStatus: {
        type: String,
        enum: ['Never', 'Former', 'Current', 'Unknown'],
        default: 'Unknown',
      },
      alcoholConsumption: {
        type: String,
        enum: ['Never', 'Occasional', 'Regular', 'Unknown'],
        default: 'Unknown',
      },
      exerciseFrequency: {
        type: String,
        enum: ['Sedentary', 'Light', 'Moderate', 'Active', 'Very Active', 'Unknown'],
        default: 'Unknown',
      },
      dietType: {
        type: String,
        enum: ['Regular', 'Vegetarian', 'Vegan', 'Pescatarian', 'Other', 'Unknown'],
        default: 'Unknown',
      },
    },

    // Insurance Information (Optional)
    insurance: {
      provider: String,
      policyNumber: String,
      expiryDate: Date,
    },
  },
  {
    timestamps: true,
  }
);

// Method to get active medications
userHealthProfileSchema.methods.getActiveMedications = function () {
  return this.currentMedications.filter((med) => med.isActive);
};

// Method to get active medical conditions
userHealthProfileSchema.methods.getActiveConditions = function () {
  return this.medicalConditions.filter((cond) => cond.isActive);
};

// Method to get primary emergency contact
userHealthProfileSchema.methods.getPrimaryEmergencyContact = function () {
  return this.emergencyContacts.find((contact) => contact.isPrimary) || this.emergencyContacts[0];
};

// Static method to get or create profile for a user
userHealthProfileSchema.statics.getOrCreate = async function (userId) {
  let profile = await this.findOne({ userId });
  if (!profile) {
    profile = await this.create({ userId });
  }
  return profile;
};

// Method to check if profile is complete
userHealthProfileSchema.methods.isComplete = function () {
  const requiredFields = [
    this.bloodGroup !== 'Unknown',
    this.emergencyContacts.length > 0,
    this.medicalConditions.length > 0 || this.allergies.length > 0,
  ];
  return requiredFields.every((field) => field);
};

module.exports = mongoose.model('UserHealthProfile', userHealthProfileSchema);
